### PR TITLE
Support "Sets" for up-to-date check inputs/outputs

### LINKED ISSUE #5596: Add a way to associate inputs and outputs in FastUpToDate
**Visual Studio Version**: 16.4 Preview 2.0

**Summary**:

Linked ASP.NET Core issue: https://github.com/aspnet/AspNetCore/issues/13204

ASP.NET Core projects produce two compiled outputs: `.cs` files (et al) produce an `App.dll`, `.cshtml` files produce an `App.Views.dll`. When a cshtml file changes, `App.dll` remains unchanged, but a newer `App.Views.dll` needs to be compiled. 

To allow a build to occur when cshtml files change, ASP.NET Core projects (via the `Microsoft.NET.Sdk.Razor` SDK), add cshtml files to `UpToCheckInput`. Fast UpToDateCheck is now aware of these files and rebuild when these files change. However doing so also leaves FastUpToDateCheck in an incosistent state - the input `.cshtml` file is always newer than the output `MyApp.dll` and consequently fast up to date check is broken until a user touches a cs file.

**Feature request**:
The feature request was to have a way to associate inputs and outputs in FUTD. By default, everything would be associated with the default `UpToDateCheckBuilt`. However, Inputs would have a way to state that they specifically affect certain outputs. When FUTD runs, it correlates these and uses this to perform the dirty check.

### PR DESCRIPTION
**Pull request by drewnoakes**
Fixes #5596.

ASP.NET's razor pages are compiled into a `*.Views.dll` assembly, which is separate from the default output assembly that other code ends up in. This can cause redundant builds in the following scenario:

1. Build the project (`MyProject.dll` and `MyProject.Views.dll` both updated)
2. Modify `Page.cshtml`
3. Build the project (`MyProject.Views.dll`  updated)
4. Build the project (incorrectly considered out-of-date because `Page.cshtml` is newer than `MyProject.dll`)

This PR provides ASP.NET (and other projects) the ability to group sets of inputs and outputs together by adding some metadata to the items.

In the ASP.NET case, you would effectively do this with:

```xml
<UpToDateCheckInput Include="Page.cshtml" Set="Views" />
<UpToDateCheckBuild Include="Path\MyProject.Views.dll" Set="Views" />
```

Build after changing `Index.cshtml` the build log would resemble (abridged):

```
1>FastUpToDate: Comparing timestamps of inputs and outputs in set 'Views': (MyProject)
1>FastUpToDate: Adding UpToDateCheckBuilt outputs in set 'Views': (MyProject)
1>FastUpToDate:     'C:\repos\MyProject\obj\Debug\netcoreapp3.0\MyProject.Views.dll' (MyProject)
1>FastUpToDate: Adding UpToDateCheckInput inputs in set 'Views': (MyProject)
1>FastUpToDate:     'C:\repos\MyProject\Pages\Page.cshtml' (MyProject)
1>FastUpToDate: Input 'C:\repos\MyProject\Pages\Page.cshtml' is newer (29/10/2019 11:00:17 PM) than earliest output 'C:\repos\MyProject\obj\Debug\netcoreapp3.0\MyProject.Views.dll' (29/10/2019 10:56:53 PM), not up to date. (MyProject)
1>FastUpToDate: Up to date check completed in 48.2 ms (MyProject)
1>------ Build started: Project: MyProject, Configuration: Debug Any CPU ------
1>MyProject -> C:\repos\MyProject\bin\Debug\netcoreapp3.0\MyProject.dll
1>MyProject -> C:\repos\MyProject\bin\Debug\netcoreapp3.0\MyProject.Views.dll
========== Build: 1 succeeded, 0 failed, 0 up-to-date, 0 skipped ==========
```

Building again immediately:

```
1>FastUpToDate: Adding UpToDateCheckBuilt outputs in set 'Views': (MyProject)
1>FastUpToDate:     'C:\repos\MyProject\obj\Debug\netcoreapp3.0\MyProject.Views.dll' (MyProject)
1>FastUpToDate: Adding UpToDateCheckInput inputs in set 'Views': (MyProject)
1>FastUpToDate:     'C:\repos\MyProject\Pages\Page.cshtml' (MyProject)
1>FastUpToDate: In set 'Views', no inputs are newer than earliest output 'C:\repos\MyProject\obj\Debug\netcoreapp3.0\MyProject.Views.dll' (29/10/2019 10:56:53 PM). (MyProject)
1>FastUpToDate: Project is up to date. (MyProject)
1>FastUpToDate: Up to date check completed in 127.7 ms (MyProject)
========== Build: 0 succeeded, 0 failed, 1 up-to-date, 0 skipped ==========
```

### CODE REVIEW

File: docs/up-to-date-check.md [hunk 1]
```diff
@@ -36,11 +39,45 @@ allows custom logic to be executed when determining the set of items. The releva
...
+
+For some advanced scenarios, it's necessary to partition inputs and outputs into groups and consider each separately.
+This can be achieved by adding `Set` metadata to the relevant items.
+
+For example, an ASP.NET project may use sets to group Razor `.cshtml` files with their output assembly `MyProject.Views.dll`,
+which is distinct from the other compilation target `MyProject.dll`. This could be achieved with something like:
+
+```xml
+<ItemGroup>
+  <UpToDateCheckInput Include="Home.csproj" Set="Views" />
```
Comment (nguerrera) [inline docs/up-to-date-check.md]: Was this meant to be Home.cshtml?

File: docs/up-to-date-check.md [hunk 2]
```diff
@@ -57,7 +60,24 @@ which is distinct from the other compilation target `MyProject.dll`. This could
...
+### Copied files
+
+Builds may copy files from a source location to a destination location. Information about these locations should be
+captured in the project so that the up-to-date check can determine if the source file is newer than the destination,
+in which case the project is out-of-date and a build will be allowed.
+
+To model this, use:
+
+```xml
+<UpToDateCheckBuilt Include="Source\File.txt" Original="Destination\File.txt" />
```
Comment (nguerrera) [inline docs/up-to-date-check.md]: Thanks for documenting this!

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs [hunk 3]
```diff
@@ -164,93 +167,112 @@ private bool CheckGlobalConditions(BuildUpToDateCheckLogger logger, State state)
...
-            DateTime? earliest = DateTime.MaxValue;
-            string? earliestPath = null;
-            bool hasOutput = false;
+            // UpToDateCheckInput/Output/Built items have optional 'Set' metadata that determine whether they
+            // are treated separately or not. If omitted, such inputs/outputs are included in the default set,
+            // which also includes other items such as project files, compilation items, analyzer references, etc.
 
-            foreach (string output in outputs)
+            // First, validate the relationship between inputs and outputs within the default set.
+            if (!CheckInputsAndOutputs(CollectDefaultInputs(), CollectDefaultOutputs(), timestampCache, DefaultSetName))
```
Comment (davidwengier) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: Why do the default set separately verses just having a special case in the log messages, which is already happening in some log messages anyway? At a quick glance it looks like you could just not remove `DefaultSetName` from `SetNames` and everything would work as expected, without needing to duplicate the CollectXX() methods?

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.Logger.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+using System;
+using System.IO;
+using Microsoft.VisualStudio.Telemetry;
+
+namespace Microsoft.VisualStudio.ProjectSystem.UpToDate
+{
+    internal sealed partial class BuildUpToDateCheck
+    {
+        private sealed class Log
```
Comment (davidwengier) [inline src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.Logger.cs]: Nit: Class is `Log`, filename is `Logger`

File: docs/up-to-date-check.md (hunk 1 above)
Comment (drewnoakes) [reply docs/up-to-date-check.md]: Yes, thanks! I have much more muscle memory for csproj :)

```suggestion
  <UpToDateCheckInput Include="Home.cshtml" Set="Views" />
```

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs (hunk 3 above)
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.cs]: The default sets contain a bunch more items that are subject to comparisons, but not defined as `BuildUpToDateCheck*` items. Have a look at the implementation of `CollectDefaultInputs`, for example.

This went through several iterations to try and remove redundancy. There may be more opportunity here, but I've stared at this for a while and can't see it.

File: docs/up-to-date-check.md [hunk 5]
```diff
@@ -36,11 +39,45 @@ allows custom logic to be executed when determining the set of items. The releva
...
+
+```xml
+<ItemGroup>
+  <UpToDateCheckInput Include="Home.cshtml" Set="Views" />
+  <UpToDateCheckOutput Include="MyProject.Views.dll" Set="Views" />
+</ItemGroup>
+```
+
+Items that do not specify a `Set` are included in the default set. Items may be added to multiple sets by separating
+their names with a semicolor (e.g. `Set="Set1;Set2"`).
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply docs/up-to-date-check.md]: Thanks, fixed.

File: docs/up-to-date-check.md [hunk 6]
```diff
@@ -36,11 +39,45 @@ allows custom logic to be executed when determining the set of items. The releva
...
+### Copied files
+
+Builds may copy files from a source location to a destination location. Information about these locations should be
+captured in the project so that the up-to-date check can determine if the source file is newer than the destination,
+in which case the project is out-of-date and a build will be allowed.
+
+To model this, use:
+
+```xml
+<UpToDateCheckBuilt Include="Source\File.txt" Original="Destination\File.txt" />
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply docs/up-to-date-check.md]: Yes it is very unintuitive. Too late to change it now unfortunately.

File: src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.ItemComparer.cs [hunk 7]
```diff
@@ -0,0 +1,29 @@
+﻿using System.Collections.Generic;
```
[the remark this answers is not in the document]
Comment (drewnoakes) [reply src/Microsoft.VisualStudio.ProjectSystem.Managed/ProjectSystem/UpToDate/BuildUpToDateCheck.ItemComparer.cs]: Thanks.

File: docs/up-to-date-check.md (hunk 6 above)
[the remark this answers is not in the document]
Comment (drewnoakes) [reply docs/up-to-date-check.md]: BTW it's been like this for at least a year, possibly quite a bit longer.

[the remark this answers is not in the document]
Comment (nguerrera) [reply docs/up-to-date-check.md]: Oh I glossed over that and was about to try this feature. I definitely would have got this backwards. So this discussion will save me some time.

**Pull request merged**